musicdao/src/main/java/com/example/musicdao/MusicService.kt

```diff
@@ -21,34 +25,81 @@ import nl.tudelft.trustchain.common.BaseActivity
 import org.apache.commons.io.FileUtils
 import java.io.*
 
-class MusicService : BaseActivity() {
+open class MusicService : BaseActivity() {
     lateinit var torrentStream: TorrentStream
     override val navigationGraph = R.navigation.musicdao_navgraph
     lateinit var contentSeeder: ContentSeeder
+    var filter: String = ""
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
-        torrentStream = TorrentStream.init(
-            TorrentOptions.Builder()
-                .saveLocation(applicationContext.cacheDir)
-                .removeFilesAfterStop(false)
-                .autoDownload(false)
-                .build())
+        torrentStream = try {
+            TorrentStream.getInstance()
+        } catch (e: Exception) {
+            TorrentStream.init(
+                TorrentOptions.Builder()
+                    .saveLocation(applicationContext.cacheDir)
+                    .removeFilesAfterStop(false)
+                    .autoDownload(false)
+                    .build())
+        }
 
         registerBlockSigner()
 
-        contentSeeder = ContentSeeder(SessionManager(), applicationContext.cacheDir)
+        iterativelyCrawlTrustChains()
+
+        contentSeeder = ContentSeeder.getInstance(SessionManager(), applicationContext.cacheDir)
         contentSeeder.start()
+
+        handleIntent(intent)
+    }
+
+    override fun onNewIntent(intent: Intent) {
+        super.onNewIntent(intent)
+        setIntent(intent)
+        handleIntent(intent)
+    }
+
+    private fun handleIntent(intent: Intent) {
+        if (Intent.ACTION_SEARCH == intent.action) {
+            intent.getStringExtra(SearchManager.QUERY)?.also { query ->
+                filter = query
+            }
+        }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        if (menu == null) return false
+        val inflater: MenuInflater = menuInflater
+        inflater.inflate(R.menu.menu_searchable, menu)
+        return true
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return when (item.itemId) {
+            R.id.action_search -> {
+                onSearchRequested()
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
+        }
     }
 
-    override fun onCreateContextMenu(
-        menu: ContextMenu?,
-        v: View?,
-        menuInfo: ContextMenu.ContextMenuInfo?
-    ) {
-        super.onCreateContextMenu(menu, v, menuInfo)
-        menuInflater.inflate(R.menu.menu_add_playlist, menu)
+    /**
+     * This is a very simplistic way to crawl all chains from the peers you know
+     */
+    private fun iterativelyCrawlTrustChains() {
+        val trustchain = IPv8Android.getInstance().getOverlay<TrustChainCommunity>()!!
+        lifecycleScope.launchWhenStarted {
+            while (isActive) {
+                for (peer in trustchain.getPeers()) {
+                    trustchain.crawlChain(peer)
```

Comment:
Here you are crawling the chains of all connected peers. This can cause issues when crawling peers with a high number of blocks. There are peers with half million tribler_bandwidth blocks which are useless for you to crawl and trying to crawl all of them at the same time could even make the app unresponsive. It would be safer and more useful to only crawl chains of peers that are useful to you. I'd recommend to create your own Community (MusicCommunity?) and only crawl chains of peers who are in that community.

Reply:
Yes you are right. I was planning to make a MusicCommunity that inherits from TrustChainCommunity, if that makes sense.

Reply:
I've never tried subclassing a TrustChainCommunity, but it should theoretically work. You are effectively creating your own chain only that is visible only by your community. One downside I see is that since you are using a separate database, you are not able to see the blocks in the TrustChain Explorer app which is showing the chain stored by TrustChainCommunity. It might not be an issue, just keep it in mind.